### PR TITLE
Make hash-breaking cleanups avoided in #26007

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

PR #26007 used these to avoid causing a mass rebuild. Now that we know things work, I'm doing a cleanup of all the individual derivations. Beyond tidiness for tidiness's sake, I think this is very important because people write packages by copying existing packages, so it's important to demonstrate the intended idem.

Some of the core infrastructure also needs a cleanup, but I'm leaving that in https://github.com/NixOS/nixpkgs/pull/26805 as it will take more effort to get correct

###### Things done

 - mass rebuild tests: http://hydra.nixos.org/jobset/nixpkgs/ericson2314-cross-staging

 - cross tests: http://hydra.nixos.org/jobset/nixpkgs/ericson2314-cross-scratch

-----------

CC @Dridus @copumpkin

### CODE REVIEW

**Comment by FRidh** [comment]
What kind of prefix is `prefix`? Maybe the attribute name could describe it better.

**Comment by Ericson2314** [comment]
@FRidh Added some comments for that.

File: pkgs/development/libraries/zlib/default.nix [hunk 1]
```diff
@@ -17,7 +16,7 @@ stdenv.mkDerivation rec {
     sha256 = "c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1";
   };
 
-  postPatch = stdenv.lib.optionalString stdenv.isDarwin ''
+  postPatch = stdenv.lib.optionalString hostPlatform.isDarwin ''
     substituteInPlace configure \
       --replace '/usr/bin/libtool' 'ar' \
```
Comment (copumpkin) [inline pkgs/development/libraries/zlib/default.nix]: An impurity like /usr/bin/ seems worth patching out regardless of the platform, right?

Comment (copumpkin) [inline pkgs/development/libraries/zlib/default.nix]: Then again, you aren't really changing this logic here, so I guess it's not worth changing here.

Comment (Ericson2314) [inline pkgs/development/libraries/zlib/default.nix]: @copumpkin you're looking at the person that just ran `s/stdenv\.is/hostPlatform.is/g' :D.

In seriousness, looks like `libtool` is being replaced with `ar`, which is kind of sketchy? I'd guess

 - libtool should continue to be used where possible

 - `ar` should be replaced with `$AR`

**Comment by Ericson2314** [comment]
Hydra seems to not want to evaluate the job?

**Comment by Ericson2314** [comment]
At least the cross tests still work, which ends up needing the guts of the mass rebuild just fine. Native coreutils has this test https://github.com/wertarbyte/coreutils/blob/master/tests/install/install-C-root fail for some reason, however.

**Comment by Ericson2314** [comment]
coreutils built after rebuild now.

**Comment by Ericson2314** [comment]
hydra.nixos.org kept on timing out, but I built the Travis commands locally on both Linux and Darwin, which should build all changed platforms. so merging to staging.

**Pull request merged**